include/bluetooth/conn.h

```diff
@@ -100,6 +100,18 @@ struct bt_conn *bt_conn_lookup_addr_le(u8_t id, const bt_addr_le_t *peer);
  */
 const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn);
 
+/** @brief Get array index of a connection
+ *
+ *  This function is used to map bt_conn to index of an array of
+ *  connections. The array has CONFIG_BT_MAX_CONN elements.
+ *
+ *  @param conn Connection object.
+ *
+ *  @return Index of the connection object.
+ *  The range of the returned value is 0..CONFIG_BT_MAX_CONN-1
+ */
+u8_t bt_conn_get_index(struct bt_conn *conn);
```

Comment:
this should be:

```
int bt_conn_get_index(struct bt_conn *conn);
```

and we should return `-ENOENT` or something if the pointer is NULL or not found in the list.

Reply:
I also agree with @jhedberg 's suggestion and use `bt_conn_index` instead.

Reply:
@carlescufi the function can't really fail if given a valid bt_conn pointer (all bt_conn pointers come from the same array). At most I'd make this an assert, since by making the return not always be an array index you're forcing callers to add complicated logic.

Reply:
The issue is that __ASSERT()s are disabled in production builds, and this could easily be a source of memory corruption. Could we have a check with `k_oops()` in case of failure? I know this is an additional branch, but still.

Reply:
If an application has a garbage bt_conn pointer this will cause bad things in any code that uses it, so I'm not sure why this API would be special? I understand the case of NULL translating into some big memory index, but if we need something that should always be there I don't think we should code it straight into this function, rather there should be a system wide "always assert" API for something like that.

Reply:
Anyway, the way to best do it (I think) is to calculate the index and then have an ALWAYS_ASSERT(id < CONFIG_BT_MAX_CONN). That should then catch both NULL and "garbage pointer" cases.